File: ts-langchain/src/profile/service.ts
/**
 * Profile Service — 用户画像 + 长期记忆 + 问答历史
 *
 * 核心能力：
 * 1. getOrCreate: 懒加载用户画像
 * 2. getRelevant: 召回相关记忆
 * 3. addMemory: 存储新记忆
 * 4. recordQA: 记录问答
 * 5. buildMemoryContext: 将记忆注入 System Prompt
 *
 * 统一通过 Repository 层访问进程内或 Cloudflare 存储后端。
 */

import {
  UserProfile,
  Memory,
  QARecord,
  createMemory,
  createQARecord,
} from "./index.js";
import { ChatOpenAI } from "@langchain/openai";
import { config } from "../config/index.js";
import { getRepositories } from "../repositories/index.js";
import type { MemorySearchResultData } from "../repositories/types.js";

// 获取当前后端的仓库实例
function getRepos() {
  return getRepositories();
}

// ============ 映射辅助函数 ==========

// 将 Gateway 画像数据映射为 UserProfile
function mapProfile(data: {
  user_id: string;
  name: string;
  preferences_json: string;
  created_at: string;
  updated_at: string;
}): UserProfile {
  return {
    id: data.user_id,
    name: data.name || "",
    preferences: JSON.parse(data.preferences_json || "{}"),
    created_at: data.created_at,
    last_active_at: data.updated_at,
  };
}

// 将 Gateway 记忆数据映射为 Memory
function mapMemory(data: {
  id: string;
  user_id: string;
  content: string;
  category: string;
  importance: number;
  created_at: string;
  updated_at: string;
}): Memory {
  return {
    id: data.id,
    user_id: data.user_id,
    content: data.content,
    category: data.category as Memory["category"],
    importance: data.importance,
    created_at: data.created_at,
    updated_at: data.updated_at,
  };
}

// 将 QARecord 序列化为字典
function qaRecordToDict(record: QARecord): Record<string, string> {
  return {
    id: record.id,
    user_id: record.user_id,
    conversation_id: record.conversation_id,
    question: record.question,
    answer: record.answer,
    timestamp: record.timestamp,
  };
}

// 将 Gateway 搜索结果映射为 Memory 列表
function mapSearchResults(
  items: MemorySearchResultData[],
  userId: string,
): Memory[] {
  return items.map((item) =>
    mapMemory({
      id: item.id,
      user_id: userId,
      content: item.content,
      category: item.category,
      importance: item.importance,
      created_at: item.created_at,
      updated_at: item.updated_at,
    }),
  );
}

// ============ Profile Service ==========

export class ProfileService {
  /**
   * 懒加载用户画像（不存在时创建）
   */
  // 获取 getOrCreate 对应的数据
  static async getOrCreate(userId: string, name: string = ""): Promise<UserProfile> {
    const repos = getRepos();
    const data = await repos.profile.getOrCreate(userId, name);
    return mapProfile(data);
  }

  /**
   * 根据用户 ID 获取已有画像，不存在时返回 undefined
   */
  // 获取 get 对应的数据
  static async get(userId: string): Promise<UserProfile | undefined> {
    const repos = getRepos();
    const data = await repos.profile.get(userId);
    return data ? mapProfile(data) : undefined;
  }

  /**
   * 更新用户画像，支持部分字段更新
   */
  // 更新或保存 update 对应的数据
  static async update(
    userId: string,
    name?: string,
    preferences?: Record<string, unknown>,
  ): Promise<UserProfile | undefined> {
    const repos = getRepos();
    const data = await repos.profile.update(userId, name, preferences);
    return data ? mapProfile(data) : undefined;
  }
}

// ============ Memory Service ==========

export class MemoryService {
  /**
   * 存储一条新的用户记忆
   */
  // 创建或注册 add 所需的数据
  static async add(
    userId: string,
    content: string,
    category: string = "fact",
    importance: number = 3,
  ): Promise<Memory> {
    await ProfileService.getOrCreate(userId);
    const repos = getRepos();
    const memoryId = crypto.randomUUID();
    const data = await repos.memory.save(
      userId,
      memoryId,
      content.trim(),
      category,
      importance,
      "user_explicit",
      undefined,
    );
    return mapMemory(data);
  }

  /**
   * 获取用户的高优先级记忆列表，按重要性排序
   */
  // 获取 getRelevant 对应的数据
  static async getRelevant(userId: string, maxItems: number = 10): Promise<Memory[]> {
    const repos = getRepos();
    const items = await repos.memory.list(userId, { limit: maxItems });
    return items.map(mapMemory).slice(0, maxItems);
  }

  /**
   * 按类别获取用户记忆
   */
  // 获取 getByCategory 对应的数据
  static async getByCategory(userId: string, category: string): Promise<Memory[]> {
    const repos = getRepos();
    const items = await repos.memory.list(userId, { category, limit: 50 });
    return items.map(mapMemory);
  }

  /**
   * 删除指定记忆，返回是否删除成功
   */
  // 删除或清理 delete 对应的数据
  static async delete(userId: string, memoryId: string): Promise<boolean> {
    const repos = getRepos();
    return repos.memory.delete(userId, memoryId);
  }

  /**
   * 获取用户的所有记忆列表
   */
  // 获取 listAll 对应的数据
  static async listAll(userId: string): Promise<Memory[]> {
    const repos = getRepos();
    const items = await repos.memory.list(userId, { limit: 50 });
    return items.map(mapMemory);
  }

  /**
   * 将记忆组装成 prompt 片段，注入 System Prompt
   */
  // 创建或注册 buildMemoryContext 所需的数据
  static async buildMemoryContext(userId: string): Promise<string> {
    const memories = await this.getRelevant(userId, 10);
    if (memories.length === 0) return "";

    const lines = ["[我记住的关于你的事]"];
    for (const m of memories) {
      lines.push(`- ${m.content}`);
    }
    lines.push("");
    return lines.join("\n");
  }

  /**
   * 从对话中提取值得记忆的事实（正则 + LLM 双层提取）
   */
  // 执行 extractMemoriesFromConversation 对应的业务逻辑
  static async extractMemoriesFromConversation(
    userId: string,
    question: string,
    answer: string,
  ): Promise<void> {
    // 第一层：正则规则立即提取
    const regexMemories = MemoryService._extractWithRegex(userId, question);
    for (const mem of regexMemories) {
      try {
        const repos = getRepos();
        await repos.memory.save(
          userId,
          mem.id,
          mem.content,
          mem.category,
          mem.importance,
          "conversation_extraction",
        );
      } catch {
        // 静默降级
      }
    }

    // 第二层：LLM 异步提取（不阻塞主流程）
    MemoryService._extractWithLLM(userId, question, answer).catch(() => {
      // LLM 提取失败静默降级
    });
  }

  /**
   * 正则规则提取 — 匹配明显的偏好/个人信息句式
   */
  // 执行 extractWithRegex 对应的业务逻辑
  private static _extractWithRegex(userId: string, question: string): Memory[] {
    const newMemories: Memory[] = [];
    const q = question.toLowerCase();

    const preferencePatterns: Array<[RegExp, string]> = [
      [/我喜欢(.+?)[。！\n]/g, "preference"],
      [/我爱(.+?)[。！\n]/g, "preference"],
      [/我讨厌(.+?)[。！\n]/g, "preference"],
      [/别(.+?)[。！\n]/g, "preference"],
      [/不要(.+?)[。！\n]/g, "preference"],
    ];

    for (const [pattern, category] of preferencePatterns) {
      const matches = q.matchAll(pattern);
      for (const match of matches) {
        const content = match[1].trim();
        if (content.length > 1 && content.length < 50) {
          newMemories.push(
            createMemory(userId, `用户喜欢/偏好: ${content}`, category, 4),
          );
        }
      }
    }

    const infoPatterns: Array<[RegExp, string]> = [
      [/我在(.+?)[。！\n]/g, "fact"],
      [/我叫(.+?)[。！\n]/g, "fact"],
      [/我是(.+?)[。！\n]/g, "fact"],
    ];

    for (const [pattern, category] of infoPatterns) {
      const matches = q.matchAll(pattern);
      for (const match of matches) {
        const content = match[1].trim();
        if (content.length > 1 && content.length < 50) {
          newMemories.push(
            createMemory(userId, `用户信息: ${content}`, category, 5),
          );
        }
      }
    }

    return newMemories;
  }

  /**
   * LLM-based 记忆提取 — 处理正则覆盖不到的复杂表达
   *
   * 通过 LLM 分析用户问题，提取值得长期记忆的事实。
   * 只提取有明确长期价值的信息（偏好、个人信息、决定），不提取临时性内容。
   */
  // 执行 extractWithLLM 对应的业务逻辑
  private static async _extractWithLLM(
    userId: string,
    question: string,
    _answer: string,
  ): Promise<void> {
    if (!config.OPENAI_API_KEY) return;

    try {
      const llm = new ChatOpenAI({
        modelName: config.OPENAI_MODEL,
        configuration: {
          baseURL: config.OPENAI_BASE_URL,
          apiKey: config.OPENAI_API_KEY,
        },
      });

      const prompt = `你是一个记忆提取助手。分析以下对话，判断是否有值得长期记住的用户信息。

规则：
1. 只提取有长期价值的信息：偏好、习惯、个人信息（职业/所在地/家庭）、重要决定
2. 不提取：临时性内容、闲聊、问候、已经知道的重复信息
3. 每条记忆控制在 30 字以内，简洁明确
4. 如果没有任何值得记住的信息，返回空数组
5. 返回 JSON 数组，每项包含 category（preference/fact/decision/context）和 content 字段

用户原文：${question}

JSON 输出（无其他内容）：`;

      const response = await llm.invoke([
        { role: "system", content: "你只输出 JSON 数组，不输出其他内容。" },
        { role: "user", content: prompt },
      ] as any);
      const text = typeof response.content === "string" ? response.content : "";

      // 解析 JSON 数组
      const jsonMatch = text.match(/\[[\s\S]*\]/);
      if (!jsonMatch) return;

      const extracted = JSON.parse(jsonMatch[0]) as Array<{
        category: string;
        content: string;
      }>;
      for (const item of extracted) {
        if (!item.content || item.content.length > 50) continue;
        try {
          const repos = getRepos();
          await repos.memory.save(
            userId,
            crypto.randomUUID(),
            item.content,
            item.category || "fact",
            3,
            "conversation_extraction",
          );
        } catch {
          // 静默降级
        }
      }
    } catch {
      // LLM 提取失败静默降级
    }
  }
}

// ============ History Service ==========

export class HistoryService {
  /**
   * 记录一条问答历史记录
   */
  // 更新或保存 record 对应的数据
  static async record(
    userId: string,
    conversationId: string,
    question: string,
    answer: string,
  ): Promise<Record<string, string>> {
    const repos = getRepos();
    const existingConversation = await repos.conversation.get(conversationId);
    if (!existingConversation) {
      await repos.conversation.create(userId, "History", "chat", conversationId);
    } else if (existingConversation.user_id !== userId) {
      throw new Error("Conversation does not belong to the requested user");
    }

    const { messages } = await repos.message.getMessages(conversationId, 200, 0);
    for (let index = messages.length - 1; index > 0; index--) {
      const assistant = messages[index];
      const user = messages[index - 1];
      if (
        assistant.role === "assistant" &&
        user.role === "user" &&
        user.content_json === question &&
        assistant.content_json === answer
      ) {
        return {
          id: assistant.id,
          user_id: userId,
          conversation_id: conversationId,
          question,
          answer,
          timestamp: assistant.created_at,
        };
      }
    }

    const timestamp = new Date().toISOString();
    const userMessageId = crypto.randomUUID();
    const assistantMessageId = crypto.randomUUID();
    await repos.message.createBatch(conversationId, userId, [
      {
        id: userMessageId,
        conversation_id: conversationId,
        user_id: userId,
        sequence_no: messages.length,
        role: "user",
        content_json: question,
        created_at: timestamp,
      },
      {
        id: assistantMessageId,
        conversation_id: conversationId,
        user_id: userId,
        sequence_no: messages.length + 1,
        role: "assistant",
        content_json: answer,
        created_at: timestamp,
      },
    ]);
    return {
      id: assistantMessageId,
      user_id: userId,
      conversation_id: conversationId,
      question,
      answer,
      timestamp,
    };
  }

  /**
   * 获取用户问答历史，支持按会话过滤
   */
  // 获取 getHistory 对应的数据
  static async getHistory(
    userId: string,
    conversationId?: string,
    limit: number = 50,
  ): Promise<Record<string, string>[]> {
    const repos = getRepos();
    const conversations = conversationId
      ? [await repos.conversation.get(conversationId)].filter(
          (conversation): conversation is NonNullable<typeof conversation> =>
            conversation !== null && conversation.user_id === userId,
        )
      : await repos.conversation.list(userId, Math.max(limit, 20), 0);
    const records: Record<string, string>[] = [];

    for (const conversation of conversations) {
      const { messages } = await repos.message.getMessages(conversation.id, 200, 0);
      let pendingQuestion: typeof messages[number] | undefined;
      for (const message of messages) {
        if (message.role === "user") {
          pendingQuestion = message;
        } else if (message.role === "assistant" && pendingQuestion) {
          records.push({
            id: message.id,
            user_id: userId,
            conversation_id: conversation.id,
            question: pendingQuestion.content_json,
            answer: message.content_json,
            timestamp: message.created_at,
          });
          pendingQuestion = undefined;
        }
      }
    }

    return records
      .sort((left, right) => right.timestamp.localeCompare(left.timestamp))
      .slice(0, limit);
  }
}

File: py-langgraph/src/profile/service.py
"""
Profile Service — 用户画像 + 长期记忆 + 问答历史

统一通过 Repository 层访问进程内或 Cloudflare 存储后端。
"""

from __future__ import annotations

import json
import re
import threading
from datetime import datetime
from typing import Any

from langchain_core.messages import HumanMessage, SystemMessage
from langchain_openai import ChatOpenAI

from src.profile.models import Memory, QARecord, UserProfile
from src.config.settings import settings
from src.repositories import get_repositories

# Repository 缓存（模块级变量，避免重复创建）
_repositories = None


# 延迟获取当前配置的仓储单例，避免模块导入阶段创建远端客户端。
def _get_repositories():
    """获取当前配置对应的仓储实例"""
    global _repositories
    if _repositories is None:
        _repositories = get_repositories()
    return _repositories


class ProfileService:
    """用户画像管理"""

    @staticmethod
    # 获取 get or create 对应的数据
    def get_or_create(user_id: str, name: str = "") -> UserProfile:
        repos = _get_repositories()
        data = repos.get_or_create_profile(user_id, name)
        return UserProfile(
            id=data["user_id"],
            name=data.get("name", ""),
            preferences=json.loads(data.get("preferences_json", "{}")),
            created_at=data.get("created_at", ""),
            last_active_at=data.get("updated_at", ""),
        )

    @staticmethod
    # 获取 get 对应的数据
    def get(user_id: str) -> UserProfile | None:
        repos = _get_repositories()
        data = repos.get_profile(user_id)
        if not data:
            return None
        return UserProfile(
            id=data["user_id"],
            name=data.get("name", ""),
            preferences=json.loads(data.get("preferences_json", "{}")),
            created_at=data.get("created_at", ""),
            last_active_at=data.get("updated_at", ""),
        )

    @staticmethod
    # 更新或保存 update 对应的数据
    def update(user_id: str, **updates) -> UserProfile | None:
        repos = _get_repositories()
        name = updates.get("name")
        preferences = updates.get("preferences")
        data = repos.update_profile(user_id, name=name, preferences=preferences)
        if not data:
            return None
        return UserProfile(
            id=data["user_id"],
            name=data.get("name", ""),
            preferences=json.loads(data.get("preferences_json", "{}")),
            created_at=data.get("created_at", ""),
            last_active_at=data.get("updated_at", ""),
        )


class MemoryService:
    """长期记忆管理"""

    @staticmethod
    # 创建或注册 add 所需的数据
    def add(user_id: str, content: str, category: str = "fact", importance: int = 3) -> Memory:
        """存储一条新记忆"""
        repos = _get_repositories()
        data = repos.save_memory(user_id, content.strip(), category, importance)
        return Memory(
            id=data["id"],
            user_id=data["user_id"],
            content=data["content"],
            category=data["category"],
            importance=data["importance"],
            created_at=data.get("created_at", ""),
            updated_at=data.get("updated_at", ""),
        )

    @staticmethod
    # 获取 get relevant 对应的数据
    def get_relevant(user_id: str, max_items: int = 10) -> list[Memory]:
        """获取用户的高优先级记忆"""
        repos = _get_repositories()
        items = repos.list_memories(user_id, limit=max_items)
        return [
            Memory(
                id=m["id"],
                user_id=user_id,
                content=m["content"],
                category=m["category"],
                importance=m["importance"],
                created_at=m.get("created_at", ""),
                updated_at=m.get("updated_at", ""),
            )
            for m in items[:max_items]
        ]

    @staticmethod
    # 获取 get by category 对应的数据
    def get_by_category(user_id: str, category: str) -> list[Memory]:
        repos = _get_repositories()
        data = repos.list_memories(user_id, category=category)
        return [
            Memory(
                id=m["id"],
                user_id=m["user_id"],
                content=m["content"],
                category=m["category"],
                importance=m["importance"],
                created_at=m.get("created_at", ""),
                updated_at=m.get("updated_at", ""),
            )
            for m in data
        ]

    @staticmethod
    # 删除或清理 delete 对应的数据
    def delete(user_id: str, memory_id: str) -> bool:
        repos = _get_repositories()
        return repos.delete_memory(user_id, memory_id)

    @staticmethod
    # 获取 list all 对应的数据
    def list_all(user_id: str) -> list[dict]:
        repos = _get_repositories()
        return repos.list_memories(user_id)

    @staticmethod
    # 创建或注册 build memory context 所需的数据
    def build_memory_context(user_id: str) -> str:
        """将记忆组装成 prompt 片段，供注入 System Prompt"""
        memories = MemoryService.get_relevant(user_id, 10)
        if not memories:
            return ""

        lines = ["[我记住的关于你的事]"]
        for m in memories:
            lines.append(f"- {m.content}")
        lines.append("")
        return "\n".join(lines)

    @staticmethod
    # 异步构建记忆上下文，允许调用方在首 token 预算内真正取消 Gateway 请求。
    async def build_memory_context_async(user_id: str) -> str:
        repos = _get_repositories()
        memories = await repos.list_memories_async(user_id, limit=10)
        if not memories:
            return ""

        lines = ["[我记住的关于你的事]"]
        for memory in memories[:10]:
            lines.append(f"- {memory['content']}")
        lines.append("")
        return "\n".join(lines)

    @staticmethod
    # 执行 extract memories from conversation 对应的业务逻辑
    def extract_memories_from_conversation(
        user_id: str, question: str, answer: str
    ) -> list[Memory]:
        """从对话中提取值得记忆的事实（正则 + LLM 双层提取）"""
        # 第一层：正则规则立即提取
        regex_memories = MemoryService._extract_with_regex(user_id, question)
        for mem in regex_memories:
            try:
                repos = _get_repositories()
                repos.save_memory(user_id, mem.content, mem.category, mem.importance)
            except Exception:
                pass

        # 第二层：LLM 异步提取（不阻塞主流程）
        if settings.memory_auto_extract:
            thread = threading.Thread(
                target=MemoryService._extract_with_llm,
                args=(user_id, question, answer),
                daemon=True,
            )
            thread.start()

        return regex_memories

    @staticmethod
    # 执行 extract with regex 对应的业务逻辑
    def _extract_with_regex(user_id: str, question: str) -> list[Memory]:
        """正则规则提取"""
        new_memories = []
        q = question.lower()

        preference_patterns = [
            (r"我喜欢(.+?)[。！\n]", "preference"),
            (r"我爱(.+?)[。！\n]", "preference"),
            (r"我讨厌(.+?)[。！\n]", "preference"),
            (r"别(.+?)[。！\n]", "preference"),
            (r"不要(.+?)[。！\n]", "preference"),
            (r"(.+?)是我喜欢的", "preference"),
        ]

        for pattern, category in preference_patterns:
            matches = re.findall(pattern, q)
            for match in matches:
                content = match.strip()
                if len(content) > 1 and len(content) < 50:
                    new_memories.append(
                        Memory(
                            id=f"mem_{datetime.now().strftime('%Y%m%d%H%M%S%f')}",
                            user_id=user_id,
                            content=f"用户喜欢/偏好: {content}",
                            category=category,
                            importance=4,
                        )
                    )

        info_patterns = [
            (r"我在(.+?)[。！\n]", "fact"),
            (r"我叫(.+?)[。！\n]", "fact"),
            (r"我是(.+?)[。！\n]", "fact"),
        ]

        for pattern, category in info_patterns:
            matches = re.findall(pattern, q)
            for match in matches:
                content = match.strip()
                if len(content) > 1 and len(content) < 50:
                    new_memories.append(
                        Memory(
                            id=f"mem_{datetime.now().strftime('%Y%m%d%H%M%S%f')}",
                            user_id=user_id,
                            content=f"用户信息: {content}",
                            category=category,
                            importance=5,
                        )
                    )

        return new_memories

    @staticmethod
    # 执行 extract with llm 对应的业务逻辑
    def _extract_with_llm(user_id: str, question: str, answer: str) -> None:
        """LLM-based 记忆提取"""
        if not settings.openai_api_key:
            return

        try:
            llm = ChatOpenAI(
                model=settings.openai_model,
                api_key=settings.openai_api_key,
                base_url=settings.openai_base_url,
            )

            prompt = (
                "你是一个记忆提取助手。分析以下对话，判断是否有值得长期记住的用户信息。\n\n"
                "规则：\n"
                "1. 只提取有长期价值的信息：偏好、习惯、个人信息（职业/所在地/家庭）、重要决定\n"
                "2. 不提取：临时性内容、闲聊、问候、已经知道的重复信息\n"
                "3. 每条记忆控制在 30 字以内，简洁明确\n"
                "4. 如果没有任何值得记住的信息，返回空数组\n"
                "5. 返回 JSON 数组，每项包含 category（preference/fact/decision/context）和 content 字段\n\n"
                f"用户问题：{question}\n"
                f"助手回答：{answer}\n\n"
                "JSON 输出（无其他内容）："
            )

            response = llm.invoke(
                [
                    SystemMessage(content="你只输出 JSON 数组，不输出其他内容。"),
                    HumanMessage(content=prompt),
                ]
            )
            text = response.content if isinstance(response.content, str) else ""

            json_match = re.search(r"\[[\s\S]*\]", text)
            if not json_match:
                return

            extracted = json.loads(json_match.group(0))
            for item in extracted:
                content = item.get("content", "").strip()
                if not content or len(content) > 50:
                    continue
                repos = _get_repositories()
                repos.save_memory(
                    user_id, content,
                    item.get("category", "fact"), 3,
                )
        except Exception:
            pass


class HistoryService:
    """问答历史管理"""

    @staticmethod
    # 更新或保存 record 对应的数据
    def record(user_id: str, conversation_id: str, question: str, answer: str) -> dict:
        repos = _get_repositories()
        conversation = repos.get_conversation(conversation_id)
        if not conversation:
            repos.create_conversation(user_id, "History", "chat", conversation_id)
        elif conversation.get("user_id") != user_id:
            raise ValueError("Conversation does not belong to the requested user")

        messages, _ = repos.get_messages(conversation_id, 200, 0)
        for index in range(len(messages) - 1, 0, -1):
            assistant = messages[index]
            user = messages[index - 1]
            if (
                assistant["role"] == "assistant"
                and user["role"] == "user"
                and user["content_json"] == question
                and assistant["content_json"] == answer
            ):
                return {
                    "id": assistant["id"],
                    "user_id": user_id,
                    "conversation_id": conversation_id,
                    "question": question,
                    "answer": answer,
                    "timestamp": assistant["created_at"],
                }

        timestamp = datetime.now().isoformat()
        user_message_id = f"msg_{datetime.now().strftime('%Y%m%d%H%M%S%f')}_user"
        assistant_message_id = f"msg_{datetime.now().strftime('%Y%m%d%H%M%S%f')}_assistant"
        repos.create_message_batch(
            conversation_id,
            user_id,
            [
                {
                    "id": user_message_id,
                    "sequence_no": len(messages),
                    "role": "user",
                    "content": question,
                    "created_at": timestamp,
                },
                {
                    "id": assistant_message_id,
                    "sequence_no": len(messages) + 1,
                    "role": "assistant",
                    "content": answer,
                    "created_at": timestamp,
                },
            ],
        )
        return {
            "id": assistant_message_id,
            "user_id": user_id,
            "conversation_id": conversation_id,
            "question": question,
            "answer": answer,
            "timestamp": timestamp,
        }

    @staticmethod
    # 获取 get history 对应的数据
    def get_history(user_id: str, conversation_id: str = None, limit: int = 50) -> list[dict]:
        repos = _get_repositories()
        if conversation_id:
            conversation = repos.get_conversation(conversation_id)
            conversations = (
                [conversation]
                if conversation and conversation.get("user_id") == user_id
                else []
            )
        else:
            conversations = repos.list_conversations(user_id, max(limit, 20), 0)

        records = []
        for conversation in conversations:
            messages, _ = repos.get_messages(conversation["id"], 200, 0)
            pending_question = None
            for message in messages:
                if message["role"] == "user":
                    pending_question = message
                elif message["role"] == "assistant" and pending_question:
                    records.append(
                        {
                            "id": message["id"],
                            "user_id": user_id,
                            "conversation_id": conversation["id"],
                            "question": pending_question["content_json"],
                            "answer": message["content_json"],
                            "timestamp": message["created_at"],
                        }
                    )
                    pending_question = None

        return sorted(records, key=lambda item: item["timestamp"], reverse=True)[:limit]
